### PR TITLE
Add JtubeSpeech Recipe

### PR DESCRIPTION
**Pull request by Takaaki-Saeki**
This PR add a TTS recipe for JtubeSpeech.
1. add recipe scripts: egs2/jtubespeech/tts1
2. edit egs2/TEMPLATE/tts1/db.sh to add "JTUBESPEECH" to the corpus list.

### CODE REVIEW

**Comment by kan-bayashi** [comment]
Thank you for great PR!
I will review it. If CI is failed, please try to fix it at first.

File: egs2/jtubespeech/tts1/local/data.sh [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    log "stage 0: Data Download"
+    local/download.sh "${db_root}"
+fi
+
+if [ ${stage} -le 0 ] && [ ${stop_stage} -ge 0 ]; then
+    log "stage 1: local/data_prep.sh"
+    # Initial normalization of the data
+    # Doesn't change sampling frequency and it's done after stages
+    local/data_prep.sh "${db_root}/jtuberaw" "${db_root}/jtubesplit" data/train "${ctcscore_pruning_threshold}"
+fi
```
Comment (kan-bayashi) [inline egs2/jtubespeech/tts1/local/data.sh:52]: Just confirmation.
Is the audio trimmed well?
If not, why don't you add silence trimming stage?
like:
https://github.com/espnet/espnet/blob/d628250ba5f1885aebe80e246c0791338f5a79e8/egs2/kss/tts1/local/data.sh#L72-L85

File: egs2/jtubespeech/tts1/local/data.sh [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+if [ ${stage} -le 2 ] && [ ${stop_stage} -ge 2 ]; then
+    log "stage 2: utils/subset_data_dir.sh"
+    # make evaluation and devlopment sets
+    utils/subset_data_dir.sh --first data/train 500 data/devtest
+    utils/subset_data_dir.sh --first data/devtest 250 data/dev
+    utils/subset_data_dir.sh --last data/devtest 250 data/test
+    n=$(( $(wc -l < data/train/wav.scp) - 500 ))
+    utils/subset_data_dir.sh --last data/train ${n} data/${train_set}
```
Comment (kan-bayashi) [inline egs2/jtubespeech/tts1/local/data.sh:61]: Is the dataset sorted by speaker?
If so, how about using randomly selected subset?
You can follow this:
https://github.com/espnet/espnet/blob/d628250ba5f1885aebe80e246c0791338f5a79e8/egs2/ruslan/tts1/local/data.sh#L60-L66

File: egs2/jtubespeech/tts1/local/data_prep.sh [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    echo "e.g.: $0 downloads/jtuberaw downloads/jtubesplit data/all -0.5"
+    exit 1
+fi
+
+set -euo pipefail
+
+# split wavfiles
+python local/split.py \
+--db_raw ${dbr} \
+--db_split ${dbs}
```
Comment (kan-bayashi) [inline egs2/jtubespeech/tts1/local/data_prep.sh:24]:
```suggestion
    --db_raw ${dbr} \
    --db_split ${dbs}
```

File: egs2/jtubespeech/tts1/local/data_prep.sh [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+# prune wavfiles with pre-computed ctcscore
+python local/prune.py \
+--score_thresh ${scorethresh} \
+--db_raw ${dbr} \
+--db_split ${dbs}
```
Comment (kan-bayashi) [inline egs2/jtubespeech/tts1/local/data_prep.sh:30]:
```suggestion
    --score_thresh ${scorethresh} \
    --db_raw ${dbr} \
    --db_split ${dbs}
```

File: egs2/jtubespeech/tts1/local/data_prep.sh [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+ROOTDIR="$(cd $(dirname $(dirname $0)); pwd)"
+
+# check arguments
+if [ $# != 4 ]; then
+    echo "Usage: $0 <db_raw> <db_split> <data_dir> <ctcscore_threshold>"
+    echo "e.g.: $0 downloads/jtuberaw downloads/jtubesplit data/all -0.5"
+    exit 1
+fi
+
+set -euo pipefail
```
Comment (kan-bayashi) [inline egs2/jtubespeech/tts1/local/data_prep.sh:19]: Maybe better to load path here.
```suggestion
set -euo pipefail
. ./path.sh
```

File: egs2/jtubespeech/tts1/local/download.sh [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    exit 1
+fi
+
+set -euo pipefail
+
+function gdrive_download () {
+  CONFIRM=$(wget --quiet --save-cookies /tmp/cookies.txt --keep-session-cookies --no-check-certificate "https://docs.google.com/uc?export=download&id=$1" -O- | sed -rn 's/.*confirm=([0-9A-Za-z_]+).*/\1\n/p')
+  wget --load-cookies /tmp/cookies.txt "https://docs.google.com/uc?export=download&confirm=$CONFIRM&id=$1" -O $2
+  rm -rf /tmp/cookies.txt
+}
```
Comment (kan-bayashi) [inline egs2/jtubespeech/tts1/local/download.sh:20]: You can use `gdown`, which is installed by default in espnet.

Comment (kan-bayashi) [reply egs2/jtubespeech/tts1/local/download.sh:20]: https://github.com/espnet/espnet/blob/3d83127d074c4af26cc86f30f0773f72a9ad5bbe/egs2/ruslan/tts1/local/data_download.sh#L23

File: egs2/jtubespeech/tts1/local/split.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+
+
+        
+
+
+
+
+
+
```
Comment (kan-bayashi) [inline egs2/jtubespeech/tts1/local/split.py:68]: Remove?

File: egs2/jtubespeech/tts1/local/subset.py [hunk 8]
```diff
@@ -0,0 +1,33 @@
+import os
```
Comment (kan-bayashi) [inline egs2/jtubespeech/tts1/local/subset.py:1]: Is this script used in the recipe?

**Comment by kan-bayashi** [comment]
Is this recipe multi-speaker? Is so, please replace the config with the following one.
https://github.com/espnet/espnet/tree/3d83127d074c4af26cc86f30f0773f72a9ad5bbe/egs2/vctk/tts1/conf/tuning
maybe only using the following configs is OK: (default is tacotron2?)
  - train_gst+xvector_conformer_fastspeech2.yaml
  - train_gst+xvector_tacotron2.yaml
  - train_gst+xvector_transformer.yaml

File: egs2/jtubespeech/tts1/run.sh [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+# 3. Phoneme + Accent
+# (e.g. k 1 0 o 1 0 k 5 -4 o 5 -4 N 5 -3 n 5 -2 i 5 -2 ch 5 -1 i 5 -1 w 5 0 a 5 0)
+# g2p=pyopenjtalk_accent
+
+# 4. Phoneme + Accent + Pause
+# (e.g. k 1 0 o 1 0 pau k 5 -4 o 5 -4 N 5 -3 n 5 -2 i 5 -2 ch 5 -1 i 5 -1 w 5 0 a 5 0)
+# g2p=pyopenjtalk_accent_with_pause
+
+./tts.sh \
```
Comment (kan-bayashi) [inline egs2/jtubespeech/tts1/run.sh:46]: if you changed the default config to the config using x-vector, please add the option `--use_xvector true` for `tts.sh`.

Comment (kan-bayashi) [reply egs2/jtubespeech/tts1/run.sh:46]: e.g.,
https://github.com/espnet/espnet/blob/3d83127d074c4af26cc86f30f0773f72a9ad5bbe/egs2/libritts/tts1/run.sh#L41

File: egs2/jtubespeech/tts1/local/data.sh (hunk 1 above)
Comment (Takaaki-Saeki) [reply egs2/jtubespeech/tts1/local/data.sh:65]: Thanks! I added the silence trimming stage.

File: egs2/jtubespeech/tts1/local/data.sh (hunk 2 above)
Comment (Takaaki-Saeki) [reply egs2/jtubespeech/tts1/local/data.sh]: Thank you. I modified data.sh as you suggested.

File: egs2/jtubespeech/tts1/run.sh (hunk 9 above)
Comment (Takaaki-Saeki) [reply egs2/jtubespeech/tts1/run.sh:46]: Thanks for your detailed instruction. I edited conf/ and run.sh for the multi-speaker setting.

File: egs2/jtubespeech/tts1/local/data.sh (hunk 1 above)
Comment (kan-bayashi) [reply egs2/jtubespeech/tts1/local/data.sh:66]: OMG This is my mistake.
I will add in another PRs.

Comment (kan-bayashi) [reply egs2/jtubespeech/tts1/local/data.sh:66]: I added it in #3467.
Sorry for bothering you.

File: egs2/jtubespeech/tts1/local/split.py (hunk 7 above)
Comment (kan-bayashi) [inline egs2/jtubespeech/tts1/local/split.py:68]:
```suggestion
                    fa.write("{} {}\n".format(stem_seg, transcript))
```

File: egs2/jtubespeech/tts1/run.sh [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    --lang jp \
+    --feats_type raw \
+    --fs "${fs}" \
+    --n_fft "${n_fft}" \
+    --n_shift "${n_shift}" \
+    --win_length "${win_length}" \
+    --use_xvector true \
```
Comment (kan-bayashi) [inline egs2/jtubespeech/tts1/run.sh:53]:
```suggestion
    --use_xvector true \
```

File: egs2/jtubespeech/tts1/local/split.py (hunk 7 above)
Comment (kan-bayashi) [inline egs2/jtubespeech/tts1/local/split.py:68]: Could you remove this blank?

File: egs2/jtubespeech/tts1/local/prune.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+
+    with open(ctcscore_path, 'r') as fr:
+        for line in fr:
+            line_list = line.strip().split()
+            dd_ctcscore[line_list[1]] += 1
+
+    for key in dd_transcript.keys():
+        assert dd_ctcscore[key] == dd_ctcscore[key]]
+    """
+
```
Comment (kan-bayashi) [inline egs2/jtubespeech/tts1/local/prune.py:82]: If not needed, could you remove it?

**Review by kan-bayashi — COMMENTED** [review]
LGTM except for minor comments.

**Comment by Takaaki-Saeki** [comment]
In the middle of Stage2+, the mfcc preparation does not work properly and the process stops with messages below.

```bash
2021-08-18T15:52:55 (tts.sh:368:main) Stage 2+: Extract X-vector: data/ -> dump/xvector (Require Kaldi)
utils/copy_data_dir.sh: copied data from data/tr_no_dev to dump/mfcc/tr_no_dev
utils/copy_data_dir.sh: file utt2dur exists in dest dump/mfcc/tr_no_dev but not in src data/tr_no_dev.  Moving it to
 ... dump/mfcc/tr_no_dev/.backup/utt2dur
utils/validate_data_dir.sh: Successfully validated data-directory dump/mfcc/tr_no_dev
utils/data/get_utt2dur.sh: working out dump/mfcc/tr_no_dev/utt2dur from dump/mfcc/tr_no_dev/segments
utils/data/get_utt2dur.sh: computed dump/mfcc/tr_no_dev/utt2dur
steps/make_mfcc.sh --nj 32 --cmd run.pl --write-utt2num-frames true --mfcc-config conf/mfcc.conf dump/mfcc/tr_no_dev
utils/validate_data_dir.sh: Successfully validated data-directory dump/mfcc/tr_no_dev
steps/make_mfcc.sh [info]: segments file exists: using that.
run.pl: 32 / 32 failed, log is in dump/mfcc/tr_no_dev/log/make_mfcc_tr_no_dev.*.log
```

The log file is as follows.

```
# extract-segments scp,p:dump/mfcc/tr_no_dev/wav.scp dump/mfcc/tr_no_dev/log/segments.1 ark:- | compute-mfcc-feats --write-utt2dur=ark,t:dump/mfcc/tr_no_dev/log/utt2dur.1 --verbose=2 --config=conf/mfcc.conf ark:- ark:- | copy-feats --compress=true --write-num-frames=ark,t:dump/mfcc/tr_no_dev/log/utt2num_frames.1 ark:- ark,scp:/workspace/espnet/egs2/jtubespeech/tts1/dump/mfcc/tr_no_dev/data/raw_mfcc_tr_no_dev.1.ark,/workspace/espnet/egs2/jtubespeech/tts1/dump/mfcc/tr_no_dev/data/raw_mfcc_tr_no_dev.1.scp 
# Started at Wed Aug 18 15:52:56 UTC 2021
#
bash: line 1: extract-segments: command not found
bash: line 1: compute-mfcc-feats: command not found
bash: line 1: copy-feats: command not found
# Accounting: time=0 threads=1
# Ended (code 127) at Wed Aug 18 15:52:56 UTC 2021, elapsed time 0 seconds
```

It seems that the commands are not loaded properly.
I'm not sure why this happens because [path.sh](http://path.sh/) is executed.
Could you please tell me how to resolve this problem?

**Comment by sw005320** [comment]
You do not seem to install Kaldi.
The Kaldi installation is optional for espnet2, but it is needed in your case.
Please check the document here https://espnet.github.io/espnet/installation.html

**Comment by Takaaki-Saeki** [comment]
Thank you for your advice. 
I have installed kaldi and confirmed that it works well up to stage6 in my local environment.

File: egs2/jtubespeech/tts1/run.sh [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+if [ "${fs}" -eq 16000 ]; then
+    # To suppress recreation, specify wav format
+    opts="--audio_format wav "
+else
+    opts="--audio_format flac "
+fi
+
+train_set=tr_no_dev
+valid_set=dev
+test_sets=test
```
Comment (kan-bayashi) [inline egs2/jtubespeech/tts1/run.sh:24]:
```suggestion
test_sets="dev test"
```

File: egs2/jtubespeech/tts1/local/data.sh [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+
+if [ -z "${JTUBESPEECH}" ]; then
+   log "Fill the value of 'JTUBESPEECH' of db.sh"
+   exit 1
+fi
+db_root=${JTUBESPEECH}
+
+train_set=tr_no_dev
+eval_set=dev
+test_set=test
```
Comment (kan-bayashi) [inline egs2/jtubespeech/tts1/local/data.sh:41]:
```suggestion
```

File: egs2/jtubespeech/tts1/local/data.sh [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+fi
+
+if [ ${stage} -le 2 ] && [ ${stop_stage} -ge 2 ]; then
+    log "stage 2: utils/subset_data_dir.sh"
+    # make evaluation and devlopment sets
+    utils/subset_data_dir.sh data/all 500 data/devtest
+    utils/subset_data_dir.sh --first data/devtest 250 data/dev
+    utils/subset_data_dir.sh --last data/devtest 250 data/test
+    n=$(( $(wc -l < data/all/wav.scp) - 500 ))
+    utils/subset_data_dir.sh --last data/all ${n} data/${train_set}
```
Comment (kan-bayashi) [inline egs2/jtubespeech/tts1/local/data.sh:75]:
```suggestion
    utils/copy_data_dir.sh data/all "data/tr_no_dev" 
    utils/filter_scp.pl --exclude data/deveval/wav.scp \ 
        data/all/wav.scp > "data/tr_no_dev/wav.scp" 
    utils/fix_data_dir.sh "data/tr_no_dev" 
```

**Comment by kan-bayashi** [comment]
Sorry, one more request.
Could you apply black for python script?
(Sorry for making some commits by myself...)

**Pull request merged**